notebooks/2018-12-14_Fortran_Files.ipynb

```diff
@@ -0,0 +1,333 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# How Fortran stores binary files"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "Fortran is still the go-to language for number crunching."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Types of Fortran Files\n",
+    "\n",
+    "There are three different native ways for Fortran to store data in files:\n",
+    "\n",
+    "1. Formatted\n",
+    "2. Unformatted\n",
+    "3. Stream\n",
+    "\n",
+    "Then, there are libraries to store the data in specific formats, for example NetCDF.\n",
+    "\n",
+    "If you want to store complex data sets for a long time, I strongly recommend NetCDF or another dedicated data format. \n",
+    "We have detailed on this blog before how to write NetCDF files with Fortran and Python, and they have features like compression and documentation that are very beneficial.\n",
+    "\n",
+    "But what if the data isn't very complex and NetCDF would be an overkill? \n",
+    "Or if you received the data from someone else and they didn't bother with this?\n",
+    "\n",
+    "This blog post will help you with your task of reading the data."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A few notes on best practices\n",
+    "\n",
+    "There are some good practices on how to deal with files in Fortran. \n",
+    "These make it easier to port data, because they will work on different systems.\n",
+    "\n",
+    "### Ensure that you know the kind of the variable\n",
+    "\n",
+    "If you write something like\n",
+    "\n",
+    "```fortran\n",
+    "integer :: ii\n",
+    "```\n",
+    "\n",
+    "you don't really know what kind of integer `ii` will be. \n",
+    "Often, you can set the default integer and real kind with compiler options, but it's far better to explicitly declare the kind in the code itself.\n",
+    "\n",
+    "Since Fortran 2003 -- and all compilers we use today are compatible with this -- you can use the intrinsic `iso_fortran_env` module to get the proper kinds:\n",
+    "\n",
+    "```fortran\n",
+    "use iso_fortran_env, only: int32, real64\n",
+    "implicit none\n",
+    "integer(kind=int32) :: ii\n",
+    "real(kind=real64) :: x(10, 100)\n",
+    "```\n",
+    "\n",
+    "In old code, you might find statements like:\n",
+    "\n",
+    "```fortran\n",
+    "integer*4 ii        ! DO NOT DO THIS\n",
+    "```\n",
+    "\n",
+    "This syntax has *never* been standard, and I strongly discourage you from using it.\n",
+    "Slightly better, but still wrong, is this:\n",
+    "\n",
+    "```fortran\n",
+    "integer(kind=4) :: ii   ! Still not good\n",
+    "```\n",
+    "\n",
+    "There is no guarantee that every compiler will use the same kind values for the same variable types.\n",
```

Comment:
This is a bit overdramatic, `kind=4` and `kind=8` will be 32 and 64 bit respectively.